irods/client_init.py

```diff
@@ -1,32 +1,68 @@
+import contextlib
+import getpass
+import os
+import sys
+
 from irods import (env_filename_from_keyword_args, derived_auth_filename)
 import irods.client_configuration as cfg
 import irods.password_obfuscation as obf
 import irods.helpers as h
-import getpass
-import os
-import sys
 
-def write_native_credentials_to_secrets_file(password, **kw):
-    env_file = env_filename_from_keyword_args(kw)
-    auth_file = derived_auth_filename(env_file)
-    old_mask = None
+@contextlib.contextmanager
+def _open_file_for_protected_contents(file_path, *arg, **kw):
+    f = old_mask = None
     try:
         old_mask = os.umask(0o77)
-        open(auth_file,'w').write(obf.encode(password))
+        f = open(file_path, *arg, **kw)
+        yield f
     finally:
         if old_mask is not None:
             os.umask(old_mask)
-            
+        if f is not None:
+            f.close()
+
+class irodsA_already_exists(Exception):
+    pass
+
+def write_native_credentials_to_secrets_file(password, overwrite = True, **kw):
+    """Write the credentials to an .irodsA file that will enable logging in with native authentication
+       using the given cleartext password.
+
+       If overwrite is False, an already existing .irodsA will not be overwritten.
+
+       Assuming no exceptions occur, this function returns True.
+    """
+    env_file = env_filename_from_keyword_args(kw)
+    auth_file = derived_auth_filename(env_file)
+    if not overwrite and os.path.exists(auth_file):
+        raise irodsA_already_exists(f'Overwriting not enabled and {auth_file} already exists.')
```

Comment:
Asking the caller to handle `Exception`s and interpret a return value feels... split. Perhaps we should consider one of these options:

1. Remove the return value for this function. If `to_encode` is not populated, we `raise` a different kind of `Exception` to match the new `Exception` being raised if the irodsA file already exists.
2. Use the return value to communicate success or failure. Rather than raising an `Exception` when an irodsA file already exists, we can use the return value to communicate this. If we want to differentiate the failure modes, we could possibly switch from a `bool` to an error code of some kind (`int`?).

I'm leaning toward option 2 myself, but I don't mind raising `Exception`s either. I just think it would be better if it was consistent and the `return` value was doing something beyond communicating success for one operation in the process. Thoughts?

Reply:
I agree. Consistency around communicating success/failure is good.

Reply:
I'm not understanding the pros/cons here.  I think it needs live discussion, or we won't have sufficient back and forth to resolve it to the team's (and my) satisfaction.

Reply:
I don't mind reverting this back to what it was, actually.  A true value means we overwrote successfully, false means we didn't.  Thoughts?

Reply:
If we want to keep it simple, I think that just returning `False` here instead of raising an `Exception` is fine. We can do a live discussion and report back here.

Reply:
Sure. We can do that today or tomorrow if that works.

Reply:
That works.

Reply:
I believe I've wrapped up @alanking's concern as to the design of the free functions' error return mechanisms, it remains only to write a test of it.

Reply:
I agree. Let's wait for his response before marking this resolved though.

Reply:
Yep, looks good. Thanks